File: backend/src/_enterprise_number.py
"""Helpers for Belgian enterprise (BCE/KBO/CBE) numbers.

An enterprise number is a 10-digit identifier whose first digit is
either ``0`` or ``1``. Inputs in the wild come with dots, spaces or
no separators at all — :func:`normalise` strips everything down to
the canonical 10-digit form, while :func:`format_human` renders it
back as ``XXXX.XXX.XXX`` for display.
"""

from __future__ import annotations

import re


_DIGITS_RE = re.compile(r"\D")


def normalise(raw: str) -> str:
    """Return the canonical 10-digit enterprise number, or raise ``ValueError``."""
    digits = _DIGITS_RE.sub("", raw)
    if len(digits) != 10 or digits[0] not in {"0", "1"}:
        raise ValueError(f"Not a valid Belgian enterprise number: {raw!r}")
    return digits


def try_normalise(raw: str) -> str | None:
    """Return the canonical number, or ``None`` if ``raw`` is not one."""
    try:
        return normalise(raw)
    except ValueError:
        return None


def format_human(number: str) -> str:
    return f"{number[0:4]}.{number[4:7]}.{number[7:10]}"

File: frontend/lib/server/enterpriseNumber.ts
/**
 * Belgian enterprise (BCE/KBO/CBE) number helpers.
 *
 * Port of backend/src/_enterprise_number.py. An enterprise number is
 * 10 digits with a leading 0 or 1; inputs in the wild use dots, spaces,
 * or nothing as separators.
 */

const NON_DIGIT_RE = /\D/g;

export function normalise(raw: string): string {
  const digits = raw.replace(NON_DIGIT_RE, "");
  if (digits.length !== 10 || (digits[0] !== "0" && digits[0] !== "1")) {
    throw new Error(`Not a valid Belgian enterprise number: ${JSON.stringify(raw)}`);
  }
  return digits;
}

export function tryNormalise(raw: string): string | null {
  try {
    return normalise(raw);
  } catch {
    return null;
  }
}

export function formatHuman(number: string): string {
  return `${number.slice(0, 4)}.${number.slice(4, 7)}.${number.slice(7, 10)}`;
}
